internal/libyaml/api_test.go

```diff
@@ -0,0 +1,461 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package libyaml
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+
+	"go.yaml.in/yaml/v4/internal/testutil/assert"
+	"go.yaml.in/yaml/v3"
+)
+
+func TestAPI(t *testing.T) {
+	RunTestCases(t, "api_test.yaml", map[string]TestHandler{
+		"api-new":       runAPINewTest,
+		"api-method":    runAPIMethodTest,
+		"api-panic":     runAPIPanicTest,
+		"api-delete":    runAPIDeleteTest,
+		"api-new-event": runAPINewEventTest,
+	})
+}
+
+func runAPINewTest(t *testing.T, tc TestCase) {
+	t.Helper()
+
+	obj := createObject(tc.Constructor)
+	runFieldChecks(t, obj, tc.Checks)
+}
+
+func runAPIMethodTest(t *testing.T, tc TestCase) {
+	t.Helper()
+
+	obj := createObject(tc.Constructor)
+
+	// Run setup if any
+	if tc.Setup != nil {
+		if setupList, ok := tc.Setup.([]interface{}); ok && len(setupList) > 0 {
+			callMethodFromList(obj, setupList, tc.Bytes)
+		}
+	}
+
+	// Call the main method
+	callMethodFromList(obj, tc.Method, tc.Bytes)
+
+	// Run checks
+	runFieldChecks(t, obj, tc.Checks)
+}
+
+func runAPIPanicTest(t *testing.T, tc TestCase) {
+	t.Helper()
+
+	obj := createObject(tc.Constructor)
+
+	// Run setup if any
+	if tc.Setup != nil {
+		if setupList, ok := tc.Setup.([]interface{}); ok && len(setupList) > 0 {
+			callMethodFromList(obj, setupList, tc.Bytes)
+		}
+	}
+
+	// The main method call should panic
+	// Want can be either a string or a single-element sequence
+	var wantMsg string
+	switch v := tc.Want.(type) {
+	case string:
+		wantMsg = v
+	case []interface{}:
+		wantMsg = v[0].(string)
+	default:
+		t.Fatalf("want must be a string or sequence, got %T", tc.Want)
+	}
+	assert.PanicMatchesf(t, wantMsg, func() {
+		callMethodFromList(obj, tc.Method, tc.Bytes)
+	}, "Expected panic: %s", wantMsg)
+}
+
+func runAPIDeleteTest(t *testing.T, tc TestCase) {
+	t.Helper()
+
+	obj := createObject(tc.Constructor)
+
+	// Run setup if any
+	if tc.Setup != nil {
+		if setupList, ok := tc.Setup.([]interface{}); ok && len(setupList) > 0 {
+			callMethodFromList(obj, setupList, tc.Bytes)
+		}
+	}
+
+	// Call Delete method
+	callMethodFromList(obj, []interface{}{"Delete"}, false)
+
+	// Run checks after delete
+	runFieldChecks(t, obj, tc.Checks)
+}
+
+func runAPINewEventTest(t *testing.T, tc TestCase) {
+	t.Helper()
+
+	event := createEventFromList(tc.Method, tc.Bytes)
+	runFieldChecks(t, &event, tc.Checks)
+}
+
+// createObject creates a Parser or Emitter based on constructor name
+func createObject(constructor string) interface{} {
+	switch constructor {
+	case "NewParser":
+		p := NewParser()
+		return &p
+	case "NewEmitter":
+		e := NewEmitter()
+		return &e
+	default:
+		panic("unknown constructor: " + constructor)
+	}
+}
```

Comment:
This is a strange way to do. Is it because you wanted to mimics Go generics?

Without generics, usually such a thing in test is done by having an interface{} field and having this field sets to NewWhatever

I feel like I don't get the purpose you are trying to achieve.

Reply:
We are using a data driven approach. We use data strings that then map to functions in Go code (or could be functions in Python code). 
The data is not Go specific. Any data string names could have been chosen here. They just happen to look the same as the Go function names in this case...